Feature/u1141 support ssl ca

mmasashi opened the pull request:
Added `set_ssl_ca` api to `Binary_log` class to support SSL connection with a custom ssl ca certificate file. 

### How to test

```
cd mysql-replication-listener
mkdir build && cd build
cmake ..
make
./tests/replaybinlog "mysql://<username>:<password>@<mysql-server>" "<ssl-ca-cert-filepath>"
```

makinada commented on include/tcp_driver.h:
```diff
@@ -156,6 +165,20 @@ class Binlog_tcp_driver : public Binary_log_driver
+    char *opt_ssl_cipher;
+    char *opt_ssl_key;
+    char *opt_ssl_crl;
+    char *opt_ssl_crlpath;
```
`m_` is a standard prefix of a member variable.  You should stick with the convertion.

makinada commented on src/tcp_driver.cpp:
```diff
@@ -387,4 +397,14 @@ bool fetch_binlog_name_and_size(MYSQL *mysql, std::map<std::string, unsigned lon
+
+int Binlog_tcp_driver::set_ssl_ca(const char *filepath)
+{
+  char *buf = new char[strlen(filepath)];
```
Memory leak, I think :)  You might want to consider using `std::string` instead of `char*` for those strings you added.  With `char*`, you have to worry about the lifespan of the instance which it points to.  With `std::string`, you don't have to since its a real object within the given scope.  I'm sure `std::string` offers type casts to `char*` which you may need.

mmasashi commented:
Thanks for the review!

- Added `m_` prefix to the member variables.
- Use std::string for ssl option member variables and argument data type of `set_ssl_ca` method.
- Since `std::string.c_str()` with an empty string returns an object, I added `opt_value` method to handle option values.

makinada commented on src/tcp_driver.cpp:
```diff
@@ -225,6 +236,14 @@ int sync_connect_and_authenticate(MYSQL *conn, const std::string &user,
   return ERR_OK;
 }
 
+const char* Binlog_tcp_driver::opt_value(const std::string *opt_str)
```
This is rather  cosmetic, but you might want to use reference instead of pointer here.
```
opt_value(const std::string &opt_str)
```
Thus, you don't need to add `&` to the argument when you call the method.  It's also consistent with the other parts of the code.

An argument passed by reference points to the same object that the caller specified.  In the sense, a reference works the same way as a pointer.

makinada commented on src/tcp_driver.cpp:
```diff
@@ -162,6 +162,17 @@ int sync_connect_and_authenticate(MYSQL *conn, const std::string &user,
+    mysql_options(conn, MYSQL_OPT_SSL_CRL, opt_value(&m_opt_ssl_crl));
+    mysql_options(conn, MYSQL_OPT_SSL_CRLPATH, opt_value(&m_opt_ssl_crlpath));
+  }
+  mysql_options(conn,MYSQL_OPT_SSL_VERIFY_SERVER_CERT, (char*)&m_opt_ssl_verify_server_cert);
```
It looks like the variable `m_opt_ssl_verify_server_cert` has never been initialized anywhere.

mmasashi replied:
I made a change to initialize it in constructor.

makinada commented:
Looks good!
